### ISSUE TITLE
symbols in dimenions

### ISSUE THREAD

**Issue by ivoysey** [ivoysey/cooltt#2]
@jonsterling can you expand on this comment a little bit? https://github.com/ivoysey/cooltt/blob/master/src/cubical/Dim.mli#L14 Is this just something that happens in NBE when you're thinking about open terms?

**Comment by jonsterling** [comment]
Yes! So what’s happening is, there is ~basically one place~ *a couple places* where we must literally compute underneath a binder: we have to go underneath the binder of coe to find out if we are coercing in a line of pi types, a line or sigma types, something else, etc.

I could not figure out a correct way to implement this except by literally generating a “fresh dimension” and running the code. So that is what this constructor is for.

**Comment by jonsterling** [comment]
Here are a few places where this is used:

1. https://github.com/ivoysey/cooltt/blob/master/src/core/Semantics.ml#L1356
2. https://github.com/ivoysey/cooltt/blob/master/src/core/Semantics.ml#L629

**Comment by ivoysey** [comment]
Gotcha; thanks! That helps.

**Comment by ivoysey** [comment]
Is this a similar sort of deal? https://github.com/ivoysey/cooltt/blob/master/src/core/DomainData.ml#L59 Or sort of the other half of the same thing?

**Comment by jonsterling** [comment]
Yes, this is the “other half” — this part is used elsewhere in Semantics.ml to optimize the substitution of an actual dimension for one of those symbols. The substitution itself is used in parts of the coe algorithm (i suggest looking for subst_dim)

**Comment by jonsterling** [comment]
and i guess that `BindSym` is also connected to this. `BindSym` is a constructor for a function out of the interval, and `do_ap` dispatches to `subst_dim` when encountering it; and `subst_dim` uses `LetSym` in its recursive calls as an optimization.

**Comment by dlicata335** [comment]
Is there anything (other than the code I mean) to read to get oriented with the way you're doing NBE?  I'm a little confused about why you would need globally gensymed symbols in addition to scoped de bruijn indices to evaluate under binders.

**Comment by jonsterling** [comment]
@dlicata335 It's an excellent question, and frankly it is something that is very hard to explain in prose. But I would suggest first attempting the following "Thought Experiment":

Suppose you have a closure `B` in the semantic domain corresponding to a line in the universe. How do you implement an operation `coe[B](V)` that would take a semantic element of `B @ 0` to a semantic element of `B @ 1`? 

Keep the following invariants in mind:

1. You are not permitted to call `quote` in the process.
2. The semantic domain must support implicit weakening: consequently, variables are represented in the domain by de bruijn levels, not indices. (This is the ancient trick of Coquand from the 90s for obtaining efficient nbe and typechecking). (This invariant is essentially the most important part of nbe algorithms --- it is roughly the reason typechecking is feasible rather than infeasible).

If you carry this above Thought experiment through, then the need for these fresh things should become clear, and the reason why it's not obvious that one can just "do it" better with de bruijn indices should also become clear. But let me know if you come up with something better than my idea! I'd love to kill these fresh names.
